internal/manifests/networking/components.go

```diff
@@ -0,0 +1,171 @@
+	}
+
+	reverse := reverseRelations(componentRelations(tempo))
+	for source, ports := range reverse {
+		for target, conn := range ports {
+			if target != componentName {
+				continue
+			}
+			peer := policyPeerFor(source, tempo)
+			np.Spec.Ingress = append(np.Spec.Ingress, networkingv1.NetworkPolicyIngressRule{
+				Ports: conn,
+				From:  []networkingv1.NetworkPolicyPeer{peer},
+			})
+		}
+	}
+
+	if len(np.Spec.Ingress) > 0 {
+		np.Spec.PolicyTypes = append(np.Spec.PolicyTypes, networkingv1.PolicyTypeIngress)
+	}
+
+	return np
+}
+
+func policyPeerFor(name string, tempo v1alpha1.TempoStack) networkingv1.NetworkPolicyPeer {
+	switch name {
+	case netPolicys3Storage, netPolicyOtelTargets:
+		return networkingv1.NetworkPolicyPeer{
+			IPBlock: &networkingv1.IPBlock{CIDR: "0.0.0.0/0"},
+		}
+	case netPolicyClusterComponents:
+		return networkingv1.NetworkPolicyPeer{
+			NamespaceSelector: &metav1.LabelSelector{},
+		}
+	default:
+		return networkingv1.NetworkPolicyPeer{
+			PodSelector: &metav1.LabelSelector{
+				MatchLabels: manifestutils.ComponentLabels(name, tempo.Name),
+			},
+		}
+	}
+}
+
+// networkRelations define connections: from -> to using NetworkPolicyPort.
+type networkRelations = map[string]map[string][]networkingv1.NetworkPolicyPort
+
+const (
+	netPolicys3Storage         = "s3"
+	netPolicyOtelTargets       = "otel"
+	netPolicyClusterComponents = "cluster"
+)
+
+func componentRelations(tempo v1alpha1.TempoStack) networkRelations {
+	var (
+		s3Conn = []networkingv1.NetworkPolicyPort{
+			{ // TODO: get this from secret?
+				Protocol: ptr.To(corev1.ProtocolTCP),
+				Port:     ptr.To(intstr.FromInt(443)),
+			},
+			{ // TODO: get this from secret?
+				Protocol: ptr.To(corev1.ProtocolTCP),
+				Port:     ptr.To(intstr.FromInt(9000)),
+			},
+		}
+		tempoGrpcConn = networkingv1.NetworkPolicyPort{
+			Protocol: ptr.To(corev1.ProtocolTCP),
+			Port:     ptr.To(intstr.FromString(manifestutils.GrpcPortName)),
+		}
+		otelHttpConn = networkingv1.NetworkPolicyPort{
+			Protocol: ptr.To(corev1.ProtocolTCP),
+			Port:     ptr.To(intstr.FromInt(4318)),
+		}
+		otelGrpcConn = networkingv1.NetworkPolicyPort{
+			Protocol: ptr.To(corev1.ProtocolTCP),
+			Port:     ptr.To(intstr.FromInt(4317)),
+		}
+	)
+	clusterIngress := map[string][]networkingv1.NetworkPolicyPort{}
+	if tempo.Spec.Template.Gateway.Enabled { // TODO: add cluster -> gateway access
+		clusterIngress[manifestutils.GatewayComponentName] = []networkingv1.NetworkPolicyPort{}
+	}
+
+	if tempo.Spec.Template.QueryFrontend.JaegerQuery.Enabled { // TODO: add cluster -> jaegerQuery access
+		clusterIngress[manifestutils.JaegerFrontendComponentName] = []networkingv1.NetworkPolicyPort{}
+	}
+	return map[string]map[string][]networkingv1.NetworkPolicyPort{
+		netPolicyClusterComponents: clusterIngress,
+		manifestutils.DistributorComponentName: {
+			manifestutils.IngesterComponentName: {
+				tempoGrpcConn,
+			},
+			netPolicyOtelTargets: {
```

Comment:
depending on how tight we want to make the policy, the distributor should only be reachable from the gateway if the gateway is enabled